typescript/phoenix-sdk/src/market.ts

```diff
@@ -0,0 +1,317 @@
+          (numQuoteTicksPerBaseUnit *
+            this.quoteLotsPerBaseUnitPerTick *
+            toNum(this.header.quoteLotSize)) /
+            quoteLotsPerBaseUnit,
+          toNum(v.numBaseLots) / this.baseLotsPerBaseUnit,
+        ]);
+      } else {
+        let prev = asks[asks.length - 1];
+        if (!prev) {
+          throw Error;
+        }
+        let numQuoteTicksPerBaseUnit = toNum(k.priceInTicks);
+        let price =
+          (numQuoteTicksPerBaseUnit *
+            this.quoteLotsPerBaseUnitPerTick *
+            toNum(this.header.quoteLotSize)) /
+          quoteLotsPerBaseUnit;
+        if (price === prev[0]) {
+          prev[1] += toNum(v.numBaseLots) / this.baseLotsPerBaseUnit;
+        } else {
+          if (asks.length == levels) {
+            break;
+          }
+          asks.push([price, toNum(v.numBaseLots) / this.baseLotsPerBaseUnit]);
+        }
+      }
+    }
+    return new Ladder(asks.reverse().slice(0, levels), bids.slice(0, levels));
+  }
+}
+
+export class Ladder {
+  bids: Array<[number, number]>;
+  asks: Array<[number, number]>;
+
+  constructor(asks: Array<[number, number]>, bids: Array<[number, number]>) {
+    this.asks = asks;
+    this.bids = bids;
+  }
+}
+
+export const deserializeMarket = (data: Buffer): Market => {
+  let offset = marketHeaderBeet.byteSize;
+  const [header] = marketHeaderBeet.deserialize(data.subarray(0, offset));
+  let remaining = data.subarray(offset);
+
+  offset = 0;
+  let baseLotsPerBaseUnit = Number(remaining.readBigUInt64LE(offset));
+  offset += 8;
+  let quoteLotsPerTick = Number(remaining.readBigUInt64LE(offset));
+  offset += 8;
+  let sequenceNumber = Number(remaining.readBigUInt64LE(offset));
+  offset += 8;
+  let takerFeeBps = Number(remaining.readBigUInt64LE(offset));
+  offset += 8;
+  let collectedAdjustedQuoteLotFees = Number(remaining.readBigUInt64LE(offset));
+  offset += 8;
+  let unclaimedAdjustedQuoteLotFees = Number(remaining.readBigUInt64LE(offset));
+  offset += 8;
+
+  remaining = remaining.subarray(offset);
+
+  let numBids = toNum(header.marketSizeParams.bidsSize);
+  let numAsks = toNum(header.marketSizeParams.asksSize);
+  let numTraders = toNum(header.marketSizeParams.numSeats);
+
+  const bidsSize =
+    16 + 16 + (16 + orderIdBeet.byteSize + restingOrderBeet.byteSize) * numBids;
+  const asksSize =
+    16 + 16 + (16 + orderIdBeet.byteSize + restingOrderBeet.byteSize) * numAsks;
+  const tradersSize =
+    16 + 16 + (16 + 32 + traderStateBeet.byteSize) * numTraders;
+  offset = 0;
+  let bidBuffer = remaining.subarray(offset, offset + bidsSize);
+  offset += bidsSize;
+  let askBuffer = remaining.subarray(offset, offset + asksSize);
+  offset += asksSize;
+  let traderBuffer = remaining.subarray(offset, offset + tradersSize);
+
+  let bidsUnsorted = deserializeTree(bidBuffer, orderIdBeet, restingOrderBeet);
+  let asksUnsorted = deserializeTree(askBuffer, orderIdBeet, restingOrderBeet);
+
+  let bids = [...bidsUnsorted].sort(
```

Comment:
does this respect time-ordering for orders at the same price level?

Reply:
I think it should

Reply:
Jk it doesn't, don't think it's that important for now, will add a TODO